rust/lance/src/dataset/transaction.rs

```diff
@@ -506,24 +521,39 @@ impl Transaction {
                 Operation::UpdateConfig { .. } => false,
                 _ => true,
             },
-            Operation::DataReplacement { .. } => match &other.operation {
+            Operation::DataReplacement { replacements } => match &other.operation {
                 Operation::Append { .. }
                 | Operation::Delete { .. }
                 | Operation::Update { .. }
                 | Operation::Merge { .. }
                 | Operation::UpdateConfig { .. } => false,
-                Operation::CreateIndex { .. } => {
-                    // TODO(rmeng): check that the new indices isn't on the column being replaced
-                    true
-                }
-                Operation::Rewrite { .. } => {
-                    // TODO(rmeng): check that the fragments being replaced are not part of the groups
-                    true
-                }
-                Operation::DataReplacement { .. } => {
-                    // TODO(rmeng): check cell conflicts
-                    true
+                Operation::CreateIndex { new_indices, .. } => {
+                    let affected_fields = new_indices
+                        .iter()
+                        .flat_map(|index| index.fields.iter())
+                        .collect::<HashSet<_>>();
+
+                    replacements.iter().any(|replacement| {
+                        replacement
+                            .1
+                            .fields
+                            .iter()
+                            .any(|field| affected_fields.contains(field))
+                    })
                 }
+                Operation::Rewrite { .. } => self.operation.modifies_same_ids(&other.operation),
+                Operation::DataReplacement {
+                    replacements: other_replacements,
+                } => replacements.iter().any(|replacement| {
+                    other_replacements.iter().any(|other_replacement| {
+                        replacement.0 == other_replacement.0
+                            && replacement
+                                .1
+                                .fields
+                                .iter()
+                                .any(|field| other_replacement.1.fields.contains(field))
```

Comment:
I think we want to make sure the conflict is at a smaller granularity, could we check and conflict only on `same_column && same_field`?